src/adapter/src/catalog.rs

```diff
@@ -158,6 +160,82 @@ pub struct CatalogPlans {
 }
 
 impl Catalog {
+    /// Initializess the `storage_controller` to understand all shards that
+    /// `self` expects to exist.
+    ///
+    /// Note that this must be done before creating/rendering collections
+    /// because the storage controller might not be aware of new system
+    /// collections created between versions.
+    async fn initialize_storage_controller_state(
+        &mut self,
+        storage_controller: &mut dyn StorageController<Timestamp = mz_repr::Timestamp>,
+        builtin_migration_metadata: BuiltinMigrationMetadata,
+    ) -> Result<(), mz_catalog::durable::CatalogError> {
+        let collections = self
+            .entries()
+            .filter(|entry| entry.item().is_storage_collection())
+            .map(|entry| entry.id())
+            .collect();
+
+        let mut state = self.state.clone();
+
+        let mut storage = self.storage().await;
+        let mut txn = storage.transaction().await?;
+
+        storage_controller
+            .initialize_state(
+                &mut txn,
+                collections,
+                builtin_migration_metadata.previous_storage_collection_ids,
+            )
+            .await
+            .map_err(mz_catalog::durable::DurableCatalogError::from)?;
+
+        // Ensure the state chagnes from initialization are visible in the
```

Comment:
```suggestion
        // Ensure the state changes from initialization are visible in the
```

Reply:
The only PR without typos is one without doc strings.